File: format_readme.js
const fs = require('fs');

const path = 'README.md';
let lines = fs.readFileSync(path, 'utf8').split(/\r?\n/);

let out = [];
let in_bash = false;
let bash_lines = [];

function process_bash_block(bash_lines) {
    const block_text = bash_lines.join('\n');
    if (block_text.includes('pipeline {') || block_text.includes('services:')) {
        return ["```bash"].concat(bash_lines).concat(["```"]);
    }
    
    let new_lines = [];
    let current_comment = [];
    
    for (let line of bash_lines) {
        let sline = line.trim();
        if (!sline) continue;
        
        if (sline.startsWith('#') && !sline.includes('!/bin/bash')) {
            current_comment.push(line);
        } else {
            let cmd = line;
            let inline_comm = "";
            let match = line.match(/^(.*?)\s+#\s+(.*)$/);
            // Don't split #!/bin/bash as it usually has a comment but it might be treated weirdly
            if (line.includes('#!/bin/bash')) {
                cmd = line.substring(0, line.indexOf('#', 1)).trimEnd();
                inline_comm = "# " + line.substring(line.indexOf('#', 1) + 1).trim();
                if (!inline_comm.includes('Shebang')) { 
                    inline_comm = "# " + "Shebang for bash scripts";
                }
            } else if (match) {
                cmd = match[1].trimEnd();
                inline_comm = "# " + match[2];
            }
            
            new_lines.push("```bash");
            if (current_comment.length > 0) {
                new_lines.push(...current_comment);
                current_comment = [];
            }
            if (inline_comm) {
                new_lines.push(inline_comm);
            }
            
            new_lines.push(cmd);
            new_lines.push("```");
            new_lines.push("");
        }
    }

    if (current_comment.length > 0) {
        new_lines.push("```bash");
        new_lines.push(...current_comment);
        new_lines.push("```");
        new_lines.push("");
    }

    return new_lines;
}

for (let line of lines) {
    if (line.trim() === '```bash') {
        in_bash = true;
        bash_lines = [];
    } else if (line.trim() === '```' && in_bash) {
        in_bash = false;
        out.push(...process_bash_block(bash_lines));
    } else if (in_bash) {
        bash_lines.push(line);
    } else {
        out.push(line);
    }
}

let cleaned_out = [];
let prev_blank = false;
for (let line of out) {
    let is_blank = (line.trim() === '');
    if (is_blank && prev_blank) continue;
    cleaned_out.push(line);
    prev_blank = is_blank;
}

fs.writeFileSync(path, cleaned_out.join('\n'), 'utf8');

File: format_readme.py
import re

with open('README.md', 'r', encoding='utf-8') as f:
    lines = f.readlines()

out = []
in_bash = False
bash_lines = []

def process_bash_block(bash_lines):
    block_text = "".join(bash_lines)
    # Skip blocks that look like scripts or Docker Compose files
    if "pipeline {" in block_text or "#!/bin/bash" in block_text or "services:" in block_text:
        return ["```bash\n"] + bash_lines + ["```\n"]
    
    new_lines = []
    current_comment = []
    
    for line in bash_lines:
        sline = line.strip()
        if not sline:
            continue
        
        if sline.startswith('#'):
            current_comment.append(line)
        else:
            m = re.search(r'^(.*?)\s+#\s+(.*)$', line)
            cmd = line
            inline_comm = ""
            if m:
                cmd = m.group(1).rstrip() + "\n"
                inline_comm = "# " + m.group(2) + "\n"
            
            new_lines.append("```bash\n")
            if current_comment:
                new_lines.extend(current_comment)
                current_comment = []
            if inline_comm:
                new_lines.append(inline_comm)
            
            new_lines.append(cmd)
            new_lines.append("```\n")
            new_lines.append("\n")

    # If trailing comments
    if current_comment:
        new_lines.append("```bash\n")
        new_lines.extend(current_comment)
        new_lines.append("```\n")
        new_lines.append("\n")

    return new_lines

for line in lines:
    if line.strip() == '```bash':
        in_bash = True
        bash_lines = []
    elif line.strip() == '```' and in_bash:
        in_bash = False
        out.extend(process_bash_block(bash_lines))
    elif in_bash:
        bash_lines.append(line)
    else:
        out.append(line)

# clean up consecutive blank lines but keep it mostly identical
cleaned_out = []
prev_blank = False
for line in out:
    is_blank = (line.strip() == '')
    if is_blank and prev_blank:
        continue
    cleaned_out.append(line)
    prev_blank = is_blank

with open('README.md', 'w', encoding='utf-8') as f:
    f.writelines(cleaned_out)
